### PR TITLE
Render introspection warnings in the schema engine

### PR DESCRIPTION
**Pull request by pimeys**
After discussing with @janpio, it's not really a breaking change if we change some of the types and forms of the warnings. There are a few new warnings for mongodb composite types. And a ton of changed tests.

Now the warnings come as nullable string to typescript. If the value is not null, the CLI should handle coloring the text and displaying it for the user. The schema engine will not add any colors.

Note: this will break CLI tests when merged. There's some work left on the CLI side until the issue is done.

Part of: https://github.com/prisma/prisma/issues/17582

### CODE REVIEW

File: schema-engine/connectors/mongodb-schema-connector/tests/introspection/remapping_names/mod.rs [hunk 1]
```diff
@@ -102,14 +101,15 @@ fn remapping_composite_fields_with_numbers() {
 
     expected.assert_eq(res.datamodel());
 
-    res.assert_warning_code(104);
+    let expect = expect![[r#"
+        *** WARNING ***
 
-    res.assert_warning("These enum values were commented out because their names are currently not supported by Prisma. Please provide valid ones that match [a-zA-Z][a-zA-Z0-9_]* using the `@map` attribute.");
```
Comment (pimeys) [inline schema-engine/connectors/mongodb-schema-connector/tests/introspection/remapping_names/mod.rs:107]: Mongodb doesn't have enums. This was always a wrong warning message.

File: schema-engine/connectors/sql-schema-connector/src/datamodel_calculator.rs [hunk 2]
```diff
@@ -16,8 +16,9 @@ pub fn calculate(schema: &sql::SqlSchema, ctx: &IntrospectionContext, search_pat
     let (schema_string, is_empty, views) = rendering::to_psl_string(&ctx);
     let warnings = warnings::generate(&ctx);
 
-    // Warning codes 5 and 6 are for Prisma 1 default reintrospection.
-    let version = if warnings.iter().any(|w| ![5, 6].contains(&w.code)) {
+    let empty_warnings = warnings.is_empty();
+
+    let version = if !empty_warnings && !warnings.uses_prisma_1_defaults() {
```
Comment (pimeys) [inline schema-engine/connectors/sql-schema-connector/src/datamodel_calculator.rs:21]: The original check was super complex to understand. So is this, but hopefully now it reads better. We still need it, remove when removing the version checker.

File: schema-engine/sql-introspection-tests/Cargo.toml [hunk 3]
```diff
@@ -17,7 +17,6 @@ connection-string.workspace = true
 pretty_assertions = "1"
 tracing-futures = "0.2"
 tokio.workspace = true
-serde_json = { version = "1.0", features = ["float_roundtrip"] }
```
Comment (pimeys) [inline schema-engine/sql-introspection-tests/Cargo.toml:20]: 👋🏾

File: schema-engine/connectors/mongodb-schema-connector/tests/introspection/dirty_data/mod.rs [hunk 4]
```diff
@@ -71,16 +70,15 @@ fn mixing_types() {
...
+        *** WARNING ***
 
-    res.assert_warning_affected(&json!([{
-        "model": "A",
-        "field": "first",
-        "tpe": "Json",
-    }]));
+        The following fields had data stored in multiple types. Either use Json or normalize data to the wanted type:
+
+          - model: A, field: first, type: Json
```
Comment (janpio) [inline schema-engine/connectors/mongodb-schema-connector/tests/introspection/dirty_data/mod.rs:78]: I think we usually render these without the empty line in between right now, and probably want to continue doing so.

Comment (pimeys) [reply schema-engine/connectors/mongodb-schema-connector/tests/introspection/dirty_data/mod.rs:78]: [`b91cf70` (#3906)](https://github.com/prisma/prisma-engines/pull/3906/commits/b91cf7049be8e82aa6916afb356e0a6d6070f96a)

File: schema-engine/sql-introspection-tests/src/lib.rs [hunk 5]
```diff
@@ -7,28 +7,6 @@ use test_setup::{BitFlags, Tags};
...
-/// be converted to JSON.
-#[macro_export]
-macro_rules! assert_eq_json {
-    ($expected:expr, $result:expr) => {
-        let val: serde_json::Value =
-            serde_json::from_str($result.as_str()).expect("The right side value was not valid JSON.");
-
-        pretty_assertions::assert_eq!($expected, val);
-    };
-}
```
Comment (tomhoule) [inline schema-engine/sql-introspection-tests/src/lib.rs:30]: aw yiss

File: schema-engine/core/src/state.rs [hunk 6]
```diff
@@ -372,15 +372,7 @@ impl GenericApi for EngineState {
...
-                            warnings: result
-                                .warnings
-                                .into_iter()
-                                .map(|warning| crate::json_rpc::types::IntrospectionWarning {
-                                    code: warning.code,
-                                    message: warning.message,
-                                    affected: warning.affected,
-                                })
-                                .collect(),
+                            warnings: result.warnings,
```
Comment (tomhoule) [inline schema-engine/core/src/state.rs:375]: nice!

File: schema-engine/connectors/sql-schema-connector/src/datamodel_calculator.rs (hunk 2 above)
Comment (tomhoule) [reply schema-engine/connectors/sql-schema-connector/src/datamodel_calculator.rs:21]: I hear P5 is coming soon

**Review by tomhoule — APPROVED** [review]
This makes things so much better. Awesome!

I haven't checked every single warning for backward compatibility and grammar and... etc. But in terms of clarity and maintainability, this is a huge win.

**Review by jkomyno — APPROVED** [review]
Nice work @pimeys. All warnings contain the same data as before. I will create an integration version of this PR and work on that TS-side.

File: schema-engine/connectors/sql-schema-connector/src/warnings/enum.rs [hunk 7]
```diff
@@ -30,25 +30,23 @@ pub(super) fn generate_warnings(r#enum: EnumPair<'_>, warnings: &mut Warnings) {
                 .map(String::from)
                 .unwrap_or_else(|| variant.name().to_string());
 
-            let warning = generators::EnumAndValue {
-                enm: r#enum.name().to_string(),
+            warnings.enum_values_with_empty_names.push(generators::EnumAndValue {
+                r#enum: r#enum.name().to_string(),
```
Comment (Druue) [inline schema-engine/connectors/sql-schema-connector/src/warnings/enum.rs:34]: ✨

File: schema-engine/connectors/schema-connector/src/warnings.rs [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+        where
+            T: fmt::Display,
+        {
+            if !items.is_empty() {
+                writeln!(f)?;
+                f.write_str(msg)?;
+                writeln!(f)?;
+
+                for item in items {
+                    writeln!(f, "  - {item}")?;
```
Comment (Druue) [inline schema-engine/connectors/schema-connector/src/warnings.rs:118]: Do we want the extra spacing here before `- ...` it looks a little weird to me combined with the newline

File: schema-engine/connectors/mongodb-schema-connector/tests/introspection/types/mod.rs [hunk 9]
```diff
@@ -442,13 +441,15 @@ fn empty_arrays() {
...
 
-    res.assert_warning_affected(&json!([{
-        "model": "A",
-        "field": "data",
-    }]));
+    let expect = expect![[r#"
+        *** WARNING ***
+
+        Could not determine the types for the following fields:
+          - model: A, field: data
```
Comment (Druue) [inline schema-engine/connectors/mongodb-schema-connector/tests/introspection/types/mod.rs:449]: Also nitpicking here; are these things we want to capitalise in our output? `Model: ...` vs `model:  ...`. I think it'd look better capitalised 😅

File: schema-engine/sql-introspection-tests/tests/views/postgresql.rs [hunk 10]
```diff
@@ -783,17 +763,14 @@ async fn re_intro_keeps_the_map(api: &mut TestApi) -> TestResult {
...
-          }
-        ]"#]];
+        *** WARNING ***
+
+        These views were enriched with `@@map` information taken from the previous Prisma schema:
+          - B
+
+        These views were enriched with `@@map` information taken from the previous Prisma schema:
+          - B
+    "#]];
```
Comment (Jolg42) [inline schema-engine/sql-introspection-tests/tests/views/postgresql.rs:773]: This is duplicate warning is unexpected to me

File: schema-engine/connectors/mongodb-schema-connector/tests/introspection/index/mod.rs [hunk 11]
```diff
@@ -1249,10 +1248,14 @@ fn unsupported_types_in_an_index() {
...
 
-    res.assert_warning_code(3);
-    res.assert_warning(
-        "These fields are not supported by the Prisma Client, because Prisma currently does not support their types.",
-    );
+    let expect = expect![[r#"
+        *** WARNING ***
+
+        These fields are not supported by the Prisma Client, because Prisma currently does not support their types:
+          - model: A, field: data, type: JavaScriptCode
```
Comment (Jolg42) [inline schema-engine/connectors/mongodb-schema-connector/tests/introspection/index/mod.rs:1255]: Not sure what is this "JavaScriptCode" 🤔

File: schema-engine/connectors/schema-connector/src/warnings.rs [hunk 12]
```diff
@@ -0,0 +1,611 @@
...
+    pub reintrospected_id_names_in_view: Vec<View>,
+    /// The field in model has a type we do not currently support in Prisma.
+    pub unsupported_types_in_model: Vec<ModelAndFieldAndType>,
+    /// The field in view has a type we do not currently support in Prisma.
+    pub unsupported_types_in_view: Vec<ViewAndFieldAndType>,
+    /// The field in view has a type we do not currently support in Prisma.
+    pub unsupported_types_in_type: Vec<TypeAndFieldAndType>,
+    /// The name of the model is taken from a previous data model.
+    pub remapped_models: Vec<Model>,
+    /// The name of the model is taken from a previous data model.
```
Comment (Jolg42) [inline schema-engine/connectors/schema-connector/src/warnings.rs:44]:
```suggestion
    /// The name of the view is taken from a previous data model.
```

File: schema-engine/connectors/mongodb-schema-connector/tests/introspection/index/mod.rs (hunk 11 above)
Comment (Jolg42) [reply schema-engine/connectors/mongodb-schema-connector/tests/introspection/index/mod.rs:1255]: Looks like it comes from Bson https://github.com/search?q=repo%3Aprisma%2Fprisma-engines%20JavaScriptCode&type=code

**Comment by Jolg42** [comment]
I think we might need to bring back the quotations`""` around the values if some values can potentially start or end by spaces

File: schema-engine/sql-introspection-tests/tests/views/postgresql.rs (hunk 10 above)
Comment (jkomyno) [reply schema-engine/sql-introspection-tests/tests/views/postgresql.rs]: Yeah, fixed

File: schema-engine/sql-introspection-tests/tests/views/postgresql.rs [hunk 13]
```diff
@@ -1209,17 +1141,17 @@ async fn reserved_name_docs_are_only_added_once(api: &mut TestApi) -> TestResult
...
+        *** WARNING ***
+
+        These views were enriched with `@@map` information taken from the previous Prisma schema:
+          - Renamedif
+<<<<<<< HEAD
+=======
+
+        These views were enriched with `@@map` information taken from the previous Prisma schema:
+          - Renamedif
+>>>>>>> f7209cce364692930db56e825b1c3c20dc04d678
```
Comment (Jolg42) [inline schema-engine/sql-introspection-tests/tests/views/postgresql.rs:1153]: ⚠️ git diff

**Review by Jolg42 — APPROVED** [review]
That's a nice cleanup 💚

File: schema-engine/sql-introspection-tests/tests/views/postgresql.rs [hunk 14]
```diff
@@ -1148,9 +1145,12 @@ async fn reserved_name_docs_are_only_added_once(api: &mut TestApi) -> TestResult
 
         These views were enriched with `@@map` information taken from the previous Prisma schema:
           - Renamedif
+<<<<<<< HEAD
+=======
 
         These views were enriched with `@@map` information taken from the previous Prisma schema:
           - Renamedif
+>>>>>>> f7209cce364692930db56e825b1c3c20dc04d678
```
Comment (janpio) [inline schema-engine/sql-introspection-tests/tests/views/postgresql.rs]: merge artifact

**Pull request merged**